src/NServiceBus.Testing/Handler.cs

```diff
@@ -280,23 +283,33 @@ public Handler<T> ExpectNotSendToSites<TMessage>(Func<TMessage, IEnumerable<stri
             throw new NotImplementedException();
         }
 
-        private static MethodInfo GetMessageHandler<TMessage>(Type targetType, Type messageType)
+        static Func<object, object, IMessageHandlerContext, Task> GetMethod(Type targetType, Type messageType, Type interfaceGenericType)
         {
-            var method = targetType.GetMethod("Handle", new[]
+            var interfaceType = interfaceGenericType.MakeGenericType(messageType);
+
+            if (!interfaceType.IsAssignableFrom(targetType))
             {
-                messageType,
-                typeof(IMessageHandlerContext)
-            });
-            if (method != null)
+                return null;
+            }
+
+            var methodInfo = targetType.GetInterfaceMap(interfaceType).TargetMethods.FirstOrDefault();
+            if (methodInfo == null)
             {
-                return method;
+                return null;
             }
 
-            var realMessageType = typeof(TMessage).IsInterface ? typeof(TMessage) : messageType;
-            var handlerType = typeof(IHandleMessages<>).MakeGenericType(realMessageType);
-            return targetType.GetInterfaceMap(handlerType)
-                .TargetMethods
-                .FirstOrDefault();
+            var target = Expression.Parameter(typeof(object));
```

Comment:
Yes of course. Sorry I sent that PR in just before going home. 

The previous code was invoking the handle method without awaiting it or calling `GetAwaiter().GetResult()`. It is possible by using weird trickery with the dynamic keyword to call the method with `GetAwaiter().GetResult()`. But I figured since we already have a battle tested implementation in the core in the `MessageHandlerRegistry` it makes sense to compile with expression trees a strongly types function delegate which allows to call the handler. The function delegate returns a `Task` and therefore you can await it or call `GetAwaiter().GetResult()`. 

The reason why you didn't see any tests failing is because almost all your handlers return `Task.FromResult(0)`. But if you had a test with for example a `Task.Delay` inside the handler what would have happened would be that you assert the validations on the handler context before the handle method is done.

I suggest we add a few more test with the following async cases?

* Use Task.Delay or Task.Yield in a handler together with a context operation.
* Do concurrent context operations in the following way

```
var t1 = context.Send();
var t2 = context.Send();
return Task.WhenAll(t1, t2);
```

// cc @timbussmann @mat-mcloughlin